### ISSUE TITLE
Improve the colorizer (IntelliSense/Semantic-based)

### ISSUE THREAD

**Issue by be5invis** [microsoft/vscode-cpptools#230]
The current built-in colorizer for C/C++ does not support:
- color operators
- color type names separately

**Comment by bobbrow** [comment]
* Also need to support colorization for inactive preprocessor blocks.

**Comment by Srekel** [comment]
@be5invis Not sure if this is what you want, but it could be a workaround for now. Since the latest VS code you can indeed set colors for operators, like so:

```
  "editor.tokenColorCustomizations": {
    "textMateRules": [
      {
        "scope": "keyword.operator",
        "settings": {
          "foreground": "#D2B17B"
        }
      }
    ]
  },
```

**Comment by Srekel** [comment]
Another thing I would like to be colorized, and I'm assuming it falls on this extension: Different colors for function _definitions_ vs function _calls_.

Like this:

![image](https://user-images.githubusercontent.com/1578924/29774732-e621f13e-8c02-11e7-8a5a-c5cd254381cb.png)

It would make it much easier to see the structure of the code at a glance.

Actually they do have slightly different TM scopes, but I tried overriding the function call's color like this but it doesn't seem to change anything:

```
  "editor.tokenColorCustomizations": {
    "textMateRules": [
      {
        "scope": "meta.function-call",
        "settings": {
          "foreground": "#D2B17B"
        }
      }
    ]
  }
```

**Comment by bobbrow** [comment]
There doesn't appear to be a way for us to do much of what we want right now because colorization is managed by TextMate grammars (regular expressions) instead of based on semantic parsing of the file.  I do see a feature request open on VS Code to enable a better mechanism https://github.com/Microsoft/vscode/issues/1967

**Comment by bobbrow** [comment]
(go and upvote that vscode issue and maybe they'll prioritize implementing it)

**Comment by DaOnlyOwner** [comment]
Any updates on this issue?

**Comment by sean-mcmanus** [comment]
@DaOnlyOwner No progress has been made in regards to improving colorization, other than general infrastructure/porting. The cpptools extension doesn't currently implement any colorization for C++. It's done by VS Code using TextMate and atom/language-c.

**Comment by bobbrow** [comment]
Did you upvote the issue I linked?  VS Code needs to give us a better API before we can make any progress on this.

**Comment by DaOnlyOwner** [comment]
What do you need to improve the colorizer? Is it enough to expose ITokenizationSupport to extensions as suggested by the author of the linked issue?

**Comment by DaOnlyOwner** [comment]
No, how do I upvote an issue?

**Comment by bobbrow** [comment]
Open the issue and click the "thumbs up" icon on the original post to add your vote.

**Comment by DaOnlyOwner** [comment]
Oh so you upvote the post not the issue itself, gotcha ;)

**Comment by bobbrow** [comment]
Microsoft/vscode#1967 was closed as a dupe of Microsoft/vscode#585, so we'll track that instead now.  Be sure to go and upvote it.

**Comment by hyeonchang** [comment]
I hope that this issue solved as soon as possible.

**Comment by bobbrow** [comment]
VSCode has a text decoration API that we intend to use for graying out inactive regions in the February update.  We're tracking that implementation with #1466.  Still need to investigate whether this will be fast enough to be used to replace the textmate coloring for the entire file or not.

**Comment by MathiasMagnus** [comment]
Indeed. Inactive regions is a welcome feature so separate Windows/Linux code paths strike out better. 

Recognized types are the most useful stuff though, so one can see immediately if a template instantiation is what it should be. I usually don't hover the mouse over variable definitions to see their exact type, but instead rely on the colorizer. If some type is not colored properly, it's either a typo or the template metaprogram failed.

**Comment by ghost** [comment]
Is there a way to disable inactive regions greying now ? Thanks

**Comment by sumonto** [comment]
Folks this is a big issue for me, 99% of my code is greyed out any workaround?

**Comment by sean-mcmanus** [comment]
@sumonto We added a setting for our next release, see https://github.com/Microsoft/vscode-cpptools/issues/1574 . https://github.com/Microsoft/vscode-cpptools/issues/1466 has a workaround (setting intelliSenseEngine to "Tag Parser" is another workaround, gray text isn't getting seen by intellisense so that code wouldn't loose any functionality that wasn't already missing).

**Comment by sumonto** [comment]
@sean-mcmanus Thanks appreciate the prompt reply

**Comment by ghost** [comment]
I'm trying to add my typedef type to the colorizer, but I'm unable to. Is there a way to add explicit type name to the colorizer ? I have tried creating a custom theme with yo code, and playing with editor.token option without success. 

The problem seems to be that the colorizer does not associate any scope to my type def, so it's impossible to colorize. 

I understand that  updating textMate live has not been tested/done yet, but if I could add explicitly my own type it would solved my problem and many other I think ;). 

I have found this : https://forum.sublimetext.com/t/colouring-typedefs-in-c/25110/10

Gauss15 talk about adding explicit type name in the C.sublime-syntax file. I have not found a similar thing in VS Code. 

Thanks for the help !

Language : C

**Comment by DollarAkshay** [comment]
I just want my class names and struct names to be colorized.

**Comment by jesta88** [comment]
This is the only reason I'm stuck with other IDEs. Uncolored types make it very difficult to read the code.

**Comment by qzxuhui** [comment]
so until now, Is there any progress on 
"highlighted the structure/class name when it be used to declare an object" ？
thanks.

**Comment by sean-mcmanus** [comment]
@qzxuhui No work has started on that yet -- it's around number 6 on our major feature TODO list.

**Comment by safield** [comment]
Here are my findings on C++ syntax highlighting...

-Adding template diamond brackets to a method call breaks syntax highlight.
```
SomeFunction(); // highlighted correctly with "Functions" token color
SomeFunction<Foo>(); // loses highlighting and is colored as default foreground.
```


-there is no way to differentiate between a method declaration/definition, and a method invocation in a color theme. This behavior is present in similar editors to VSCode. In the following snippet, both "Foo" strings will be colored the same, where as in Sublime Text, you can theme them separately.

```
void MyClass::Foo()
{
    Foo();
}
```


Class definitions are broken. In the following snippet, everything after the word "class" is defaulted to the same standard foreground color. Expected behavior would be the public keyword receives keyword color and the class declaration receives a unique color as well.

```
class NameSpace::SomeClass: public NameSpace2::SomeOtherClass
{
};
```

**Comment by lonelyion** [comment]
I wish to have C++ STL types (like vector, string, list, etc) colorized like CodeBlocks
![image](https://user-images.githubusercontent.com/25267689/48140029-e5c24300-e2e2-11e8-8f1d-1cfc58f65e6b.png)

Currently STL types are not highlighted as if they are plain text in VSCode

**Comment by red-rj** [comment]
something that's been bothering me is `static_assert`, it should be colored like a keyword

vscode:
![colorvscode](https://user-images.githubusercontent.com/25366969/49021860-bb91d180-f17a-11e8-8326-6ca4354e0104.PNG)
visual studio:
![colorvs](https://user-images.githubusercontent.com/25366969/49021875-c2204900-f17a-11e8-836c-7df2d21e2ada.PNG)

**Comment by sean-mcmanus** [comment]
@rehdi93 This is being tracked by https://github.com/atom/language-c/issues/246 . You could upvote that issue.

**Comment by tanveerbadar** [comment]
virtual base classes/structs are not correctly colorized if you have a definition like:

class B : virtual public A
{
};

**Comment by JasonLearning** [comment]
Can class names and struct names to be colorized yet?And I'm interested what's the first one in TODO list.

**Comment by sean-mcmanus** [comment]
@JasonLearning Not yet. We're working on improving the configuration experience, performance, and build tasks.

**Comment by NtReleaseEpinephrine** [comment]
It's been several years now and VS Code still does not support proper syntax highlighting for C and C++.

Now, I'm not saying it is the fault of anyone here, but proper syntax highlighting for C and C++ is something that should have been implemented into VS Code and properly tested by test engineers before VS Code was even dropped on GitHub. 

Microsoft own VS Code and they own the C and C++ extension. There should be good communication between the people working on both projects to ensure that the people working on this extension get what they need to improve what they already have and ensure that the users of the extension get what they need for its full potential to be leveraged.

Microsoft have billions upon billions upon billions of dollars, contracts with numerous government agencies and world-wide OS, Office suite and IDE market-share domination - not to mention being one of the leaders in world-wide web browser market-share domination from Internet Explorer and Microsoft Edge - and they still cannot handle something like a proper C and C++ syntax highlighter for VS Code.

It's fascinating.

**Comment by bobbrow** [comment]
Thank you for self-moderating your comment.  I empathize with your frustration, and I'm sorry we haven't been able to address this issue sooner.  Our team has to make difficult decisions about what to work on next given the funding we receive and the data we get through this channel (GitHub) and our surveys.  This issue is high on our list and we plan to assign it to an active milestone soon.

**Comment by jesta88** [comment]
Thanks for the update. It's the only thing that prevents us switching to vscode at the moment. It's a pretty big deal for C/C++ programmers.

**Comment by sean-mcmanus** [comment]
@NtReleaseEpinephrine What exactly do you mean by "proper syntax highlighting"? VS Code shipped with a TextMate grammar for doing syntax colorization (with our extension doing none). It recently switched to a new fork of the TextMate grammar at https://github.com/jeff-hykin/cpp-textmate-grammar. So VS Code Insiders has more colorization fixes than VS Code, and using the Better C++ Syntax extension will get even more recent fixes.

Bob is talking about integrating Visual Studio C++'s syntax colorization for proper colorization of user defined types, similar to how we colorize inactive regions as transparent.

**Comment by MathiasMagnus** [comment]
@sean-mcmanus I think we're on the same page. I guess when people refer to "proper syntax highlight", they mean coloring guided by IntelliSense, that is including user defined types, preprocessor region understanding and all that jazz. The things a C++ dev might have gotten used to in big brother VS.

**Comment by Lennon925** [comment]
![123](https://user-images.githubusercontent.com/39362073/55713953-4eddec00-5a24-11e9-9c50-23aad1caed2b.PNG)
![456](https://user-images.githubusercontent.com/39362073/55713957-51404600-5a24-11e9-8e1e-63d6b7e6ae21.PNG)

I defined a clas and called this class in another file but the class name color is white, it isn't colored.
this issue can be solved? I think it's so  inconvenient for c/c++ developers.

**Comment by hackershi** [comment]
function call colorizaiton and stl type(string, vector, set, etc) colorizaiton is broken.

**Comment by mixandmatch025** [comment]
colorizing structs and macros would be really helpful

**Comment by sborkowski** [comment]
I too would like to see improved syntax highlighting. My biggest annoyance is that defines, macros and typedefs are not colorized when used, and they should be. Typedefs, at the very least, should use the same colors as built in types (int, char).

**Comment by Wiz1999** [comment]
From @safield:
> -there is no way to differentiate between a method declaration/definition, and a method invocation in a color theme. This behavior is present in similar editors to VSCode. In the following snippet, both "Foo" strings will be colored the same, where as in Sublime Text, you can theme them separately.
> 
> ```
> void MyClass::Foo()
> {
>     Foo();
> }
> ```

Just switching over to using VSCode and I've noticed this as well.  Coming from a very old editor that handled colorization of method declarations separately and struggling without this support.  Is there any outlook to when this can be fixed?  Or is there a workaround/solution we can try?  Thanks.

**Comment by qzxuhui** [comment]
If the member and member function can be highlighted in different color, it would be great help.

**Comment by Lennon925** [comment]
@sean-mcmanus I saw you modified the milestones to June 2019, So this will be supported at the end of June?

**Comment by bobbrow** [comment]
That's the goal.  We're still working with the VS Code team to provide us a proper API for this, but in the meantime we've figured out a way to get the colors in the editor without it and the performance looks acceptable.  We will also provide a setting to disable the experimental coloring in the event that anyone has performance problems or prefers the TextMate coloring.

**Comment by Lennon925** [comment]
> That's the goal. We're still working with the VS Code team to provide us a proper API for this, but in the meantime we've figured out a way to get the colors in the editor without it and the performance looks acceptable. We will also provide a setting to disable the experimental coloring in the event that anyone has performance problems or prefers the TextMate coloring.

It sounds good. Thanks for your reply.

**Comment by Lennon925** [comment]
> That's the goal. We're still working with the VS Code team to provide us a proper API for this, but in the meantime we've figured out a way to get the colors in the editor without it and the performance looks acceptable. We will also provide a setting to disable the experimental coloring in the event that anyone has performance problems or prefers the TextMate coloring.

Is it possible to set different colors for macro/struct/Union/enum variables and local/global variables if semantic-based finished?

**Comment by aminukano585** [comment]
> That's the goal. We're still working with the VS Code team to provide us a proper API for this, but in the meantime we've figured out a way to get the colors in the editor without it and the performance looks acceptable. We will also provide a setting to disable the experimental coloring in the event that anyone has performance problems or prefers the TextMate coloring.

Nice to hear this, so we should expect the rollout in a months time or less right?

**Comment by bobbrow** [comment]
@Lennon925 you can customize the colors using the `"editor.tokenColorCustomizations"` setting.  I'm not seeing a good reference online on all the scopes available, so we might need to write up some documentation on this ourselves.  Not all themes provide colors for all of the scopes either, so we expect people to need to make some customizations to get the exact colorization they want.

@aminukano585 we are currently testing the implementation and doing some optimizations.  We're hoping to have something available for insiders to try in about 2 weeks or so.

**Comment by Lennon925** [comment]
@bobbrow 

It can't solve my problem to customize the colors using the 'editor.tokenColorCustomizations' setting. Because it can't distinguish variables type. It colorize ENUM/Struct/Union/Macro with the same color.

**Comment by sean-mcmanus** [comment]
@Lennon925 We support local/global var colors with variable.other.local and variable.other.global, and macros with entity.name.function.preprocessor, but enum/struct/union appear to just map to entity.name.type -- @Colengms would it be easy/possible to distinguish between enum/struct/unions?

**Comment by Lennon925** [comment]
> @Lennon925 We support local/global var colors with variable.other.local and variable.other.global, and macros with entity.name.function.preprocessor, but enum/struct/union appear to just map to entity.name.type -- @Colengms would it be easy/possible to distinguish between enum/struct/unions?

It seems that it works just for definitions. For references, it doesn't work.
![3](https://user-images.githubusercontent.com/39362073/58363753-dd76d180-7edb-11e9-8ea1-28dd9a6fb597.PNG)

**Comment by Lennon925** [comment]
> @Lennon925 you can customize the colors using the `"editor.tokenColorCustomizations"` setting. I'm not seeing a good reference online on all the scopes available, so we might need to write up some documentation on this ourselves. Not all themes provide colors for all of the scopes either, so we expect people to need to make some customizations to get the exact colorization they want.

Link below maybe help you.
https://www.sublimetext.com/docs/3/scope_naming.html#variable

**Comment by sean-mcmanus** [comment]
@Lennon925 We haven't shipped the feature yet -- if you're using the in-progress TypeScript branch, it won't work due to the binaries not being published yet -- I think we'll need to publish the pre-release binaries when the change is ready to be checked in in order for the GitHub tests to pass.

UPDATE: Fixed via https://github.com/microsoft/vscode-cpptools/pull/3651 and https://github.com/microsoft/vscode-cpptools/pull/3738 .

**Comment by bobbrow** [comment]
We have released an [insiders build](https://github.com/microsoft/vscode-cpptools/releases/tag/0.24.0-insiders3) with the colorizer improvements.

We would encourage you to try out the feature and provide feedback on the experience.  Please note that many VS Code themes don't have full support for all of the code items (yet), so you might need to augment your themes with customizations.  There is more information, including full example settings for light/dark themes here: [Colorization docs](https://github.com/microsoft/vscode-cpptools/blob/master/Documentation/LanguageServer/colorization.md)

**Comment by VesCodes** [comment]
Really appreciate we have typedef colourisation now! However one issue that introduces is that it seems to always take precedence over whatever the original scope colourisation was? In this example case `AClass` gets scoped as an `entity.name.type`, which it is, BUT the original scope for this scenario is `entity.name.scope-resolution` which I'd rather it retain so it retains the same colouration as `::AMethod`.
![image](https://user-images.githubusercontent.com/1884844/60757020-d7a4fd80-9ffc-11e9-893b-eb7c4b0c77d8.png)

Also would be nice if non-function macros could get a separate scope so defines used as variables wouldn't get colourised the same as defines used as functions.

**Comment by marcmerlin** [comment]
So, I know it's bad form to hijack a bug, but hopefully this enough of the same issue to be ok here?
From this bug, ifdefs should now be parsed for colorization, is that also true for includes?
In this case I'm getting a warning for an include I don't have but don't care about since the ifdef does not apply to me.
Is that the same issue and it should be fixed?
Is that a new issue that needs a bug?
Or is that an existing issue with an existing bug I wasn't able to find?
![image](https://user-images.githubusercontent.com/1369412/81527514-52226600-930f-11ea-8a3d-131d2b8f6582.png)

Thanks.

**Comment by JimmyDeemo** [comment]
> Really appreciate we have typedef colourisation now!

Please can you tell me how to access this? Or does it not work with objective-c files?

<img width="267" alt="image" src="https://user-images.githubusercontent.com/4234718/93002991-cde92500-f532-11ea-960e-284a84948774.png">

**Comment by sean-mcmanus** [comment]
@JimmyDeemo Our extension doesn't run on Objective-C/C++ files. Our parser isn't configured to parse it correctly, so semantic colorization wouldn't be correct.

**Comment by JimmyDeemo** [comment]
Thanks for explaining. 👍
